### PR TITLE
add workflow for pushing nix flake artifacts to Cachix

### LINKED ISSUE #1717: Nix binary cache for flake
This project is a Nix flake. There's a service called [Cachix][1] that lets you set up a binary cache to store CI builds, so that Nix users can download binaries for the master branch instead of having to build it locally.

Cachix has [very good GitHub Actions integration][2] and is free for open source projects. Having this set up would be helpful for Nix users and take nearly no effort to maintain.

This needs to be done by the maintainer because it requires secrets to be generated and set for the repository.

[1]: https://docs.cachix.org/getting-started
[2]: https://nix.dev/tutorials/continuous-integration-github-actions

### PR DESCRIPTION
**Pull request by the-mikedavis**
closes #1717 

based on https://github.com/cachix/install-nix-action/tree/b2b9fc6cf9ad2446d49c06143ad34b98bf73be99#usage-with-flakes

I grabbed the `helix` cache on Cachix. I'll need to coordinate with @archseer to get the auth token secret set up in actions.

### CODE REVIEW

**Comment by the-mikedavis** [comment]
(WIP - will add some docs with instructions for using the cache)

**Comment by colemickens** [comment]
❤️ 🎊 looks great!

File: .github/workflows/cachix.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+jobs:
+  publish:
+    name: Publish Flake
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout sources
+      uses: actions/checkout@v2
+      with:
+        submodules: true
```
Comment (colemickens) [inline .github/workflows/cachix.yml:14]: Do you 100% need this? IIRC it looked like the flake hardcodes a snapshot of the grammars anyway.

**Comment by archseer** [comment]
Added the `CACHIX_AUTH_TOKEN` secret.

File: .github/workflows/cachix.yml (hunk 1 above)
Comment (the-mikedavis) [reply .github/workflows/cachix.yml:14]: ah good call, probably don't need this :+1:

**Comment by the-mikedavis** [comment]
I wonder if this should be enabled on push to master plus pull_request? That might make the cache fill up quickly though :thinking:

**Review by archseer — APPROVED** [review]

**Comment by anna328p** [comment]
> I wonder if this should be enabled on push to master plus pull_request? That might make the cache fill up quickly though thinking

Cachix does automatic garbage collection

**Comment by the-mikedavis** [comment]
Even with the garbage collection I'm worried the 5GB would fill up too fast with pull_request on. If we have a lot of PRs open and master isn't updated in a while, the master build might be pruned to make space for builds on PRs, right?

OTOH I think it would be useful to cache pull_requests - it would make it very easy to jump between branches when reviewing or trying out new changes.

**Pull request merged**